File: 3BV2/PP/P1/Ejercicios/04-Lab5/Java/Ferry.java
import java.util.*;

public class Main {
    // Define la capacidad máxima de los ferris
    static final int CAPACITY = 10;
    // Crea una pila (Stack) para el ferris Portos
    static Stack<String> portos = new Stack<>();
    // Crea una cola (Queue) para el ferris Celebrian
    static Queue<String> celebrian = new LinkedList<>();

    public static void main(String[] args) {
        // Crea un objeto Scanner para leer la entrada del usuario
        Scanner scanner = new Scanner(System.in);

        // Bucle infinito hasta que el programa se cierre manualmente
        while (true) {
            // Solicita al usuario la matrícula del auto
            System.out.println("Introduce la matrícula del auto:");
            String licensePlate = scanner.nextLine();

            // Solicita al usuario en qué ferris irá el auto
            System.out.println("¿En qué ferris irá el auto? (P o C)");
            String ferry = scanner.nextLine();

            // Si el usuario eligió Portos y hay espacio, agrega el auto a Portos
            if (ferry.equalsIgnoreCase("P")) {
                if (portos.size() < CAPACITY) {
                    portos.push(licensePlate);
                    System.out.println("Auto agregado a Portos.");
                } else {
                    System.out.println("Lo sentimos, Portos está lleno.");
                }
            // Si el usuario eligió Celebrian y hay espacio, agrega el auto a Celebrian
            } else if (ferry.equalsIgnoreCase("C")) {
                if (celebrian.size() < CAPACITY) {
                    celebrian.add(licensePlate);
                    System.out.println("Auto agregado a Celebrian.");
                } else {
                    System.out.println("Lo sentimos, Celebrian está lleno.");
                }
            // Si el usuario no eligió P ni C, muestra un mensaje de error
            } else {
                System.out.println("Opción no válida.");
            }

            // Muestra el contenido de ambos ferris
            printContents();

           // Si ambos ferris están llenos, muestra un mensaje y vacía ambos ferris
            if (portos.size() >= CAPACITY && celebrian.size() >= CAPACITY) {                
                System.out.println("\n=====================================================================");
                System.out.println("Los ferris han partido. Los autos no pueden ingresar.");
                System.out.println("Los ferris han completado su viaje. Los autos comenzarán a salir.");
                System.out.println("=====================================================================\n");
                unloadFerries();
                return;
            }
        }
    }

    // Método que imprime el contenido de ambos ferris
    public static void printContents() {
        System.out.println("Contenido de Portos: " + portos);
        System.out.println("Contenido de Celebrian: " + celebrian);
    }

    // Método que simula un viaje vaciando ambos ferris
    public static void unloadFerries() {
        // Vacia el ferris Portos
        while (!portos.isEmpty()) {
            String pop = portos.pop();
            System.out.println("Auto " + pop + " ha salido de Portos.");
            printContents();
        }
        
        // Vacia el ferris Celebrian
        while (!celebrian.isEmpty()) {
             String remove = celebrian.remove();
            System.out.println("Auto " + remove + " ha salido de Celebrian.");
            printContents();
        }
    }
}

File: 3BV2/PP/P1/Ejercicios/04-Lab5/Python/ferry.py
from collections import deque

# Capacidad máxima de los ferris
CAPACIDAD_MAXIMA = 10

# Inicializar los ferris
Portos = []  # Pila
Celebrian = deque()  # Cola

while len(Portos) < CAPACIDAD_MAXIMA or len(Celebrian) < CAPACIDAD_MAXIMA:
    # Leer matrícula y ferris
    matricula = input("Ingrese la matrícula del auto: ")
    ferris = input("Ingrese el ferris donde irá el auto (P para Portos, C para Celebrian): ")

    # Validar entrada y agregar auto al ferris correspondiente
    if ferris == 'P' and len(Portos) < CAPACIDAD_MAXIMA:
        Portos.append(matricula)
    elif ferris == 'C' and len(Celebrian) < CAPACIDAD_MAXIMA:
        Celebrian.append(matricula)
    else:
        print("El ferris está lleno o la opción ingresada es incorrecta. Por favor, intente nuevamente.")

    # Imprimir el contenido de los ferris
    print("Portos: ", Portos)
    print("Celebrian: ", list(Celebrian))

# Simular que inicia el viaje
print("\n=====================================================================")
print("Los ferris han partido. Los autos no pueden ingresar.")
# Simular el viaje de los ferris
print("Los ferris han completado su viaje. Los autos comenzarán a salir.")
print("=====================================================================\n")

    # Simular que los autos salen de los ferris Portos y Celebrian
while Portos or Celebrian:
    # Sacar auto de Portos
    if Portos:
        # mostrar el auto que sale de Portos
        print(f"\nEl auto {Portos[-1]} ha salido de Portos.")
        # sacar el auto de Portos
        Portos.pop() 
        # mostrar los autos restantes en Portos
        print(f"Autos restantes en Portos: {Portos}")
    
    # Sacar auto de Celebrian
    if Celebrian:
        # mostrar el auto que sale de Celebrian
        print(f"\nEl auto {Celebrian[0]} ha salido de Celebrian.")
        # sacar el auto de Celebrian
        Celebrian.popleft()
        # mostrar los autos restantes en Celebrian
        print(f"Autos restantes en Celebrian: {list(Celebrian)}")
